### ISSUE TITLE
token CA hash error when using a system-ca cert that is registered as a system root

### ISSUE THREAD

**Issue by jandubois** [k3s-io/k3s#2731]
**Environmental Info:**
K3s Version:
v1.19.5+k3s2 

Node(s) CPU architecture, OS, and Version:
Linux minikube 4.19.157 #1 SMP Tue Dec 8 13:37:29 PST 2020 x86_64 GNU/Linux

Cluster Configuration:
single node server

**Describe the bug:**
Running inside minikube, attempting to use the pre-generated minikube CA:

```
rm -rf /var/lib/k3s/
mkdir -p /var/lib/k3s/server/tls
cp /var/lib/minikube/certs/ca.crt /var/lib/k3s/server/tls/server-ca.crt
cp /var/lib/minikube/certs/ca.key /var/lib/k3s/server/tls/server-ca.key
sudo /usr/local/bin/k3s server --data-dir /var/lib/k3s --write-kubeconfig /var/lib/minikube/kubeconfig --https-listen-port 8443 --debug
```

This fails with

```
INFO[2020-12-19T02:14:14.068345432Z] Running kube-controller-manager --address=127.0.0.1 --allocate-node-cidrs=true --bind-address=127.0.0.1 --cluster-cidr=10.42.0.0/16 --cluster-signing-cert-file=/var/lib/k3s/server/tls/client-ca.crt --cluster-signing-key-file=/var/lib/k3s/server/tls/client-ca.key --kubeconfig=/var/lib/k3s/server/cred/controller.kubeconfig --leader-elect=false --port=10252 --profiling=false --root-ca-file=/var/lib/k3s/server/tls/server-ca.crt --secure-port=0 --service-account-private-key-file=/var/lib/k3s/server/tls/service.key --use-service-account-credentials=true
INFO[2020-12-19T02:14:14.069730659Z] Node token is available at /var/lib/k3s/server/token
INFO[2020-12-19T02:14:14.069787408Z] To join node to cluster: k3s agent -s https://192.168.64.63:8443 -t ${NODE_TOKEN}
INFO[2020-12-19T02:14:14.070788898Z] Wrote kubeconfig /var/lib/minikube/kubeconfig
INFO[2020-12-19T02:14:14.070873855Z] Run: k3s kubectl
INFO[2020-12-19T02:14:14.071017445Z] Module overlay was already loaded
INFO[2020-12-19T02:14:14.071119039Z] Module nf_conntrack was already loaded
INFO[2020-12-19T02:14:14.071133686Z] Module br_netfilter was already loaded
INFO[2020-12-19T02:14:14.071146430Z] Module iptable_nat was already loaded
INFO[2020-12-19T02:14:14.071603155Z] Waiting for API server to become available
ERRO[2020-12-19T02:14:14.142084527Z] token CA hash does not match the server CA hash: d5f671d11474cf600edf08de4bb1e9ecdfb98e5d14ddf97b6033d8aa5376dd4f != e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

The problem appears to be that the same CA cert is also installed in `/usr/share/ca-certificates`:

```
$ ls -l /usr/share/ca-certificates
total 4
-rw-r--r-- 1 root root 1111 Dec 18 07:48 minikubeCA.pem
drwxr-xr-x 2 root root 2600 Dec  8 21:24 mozilla
```

When I remove `minikubeCA.pem` from the system store the token error goes away. When I copy the cert back, the error can be reproduced again.

@brandond suspects that the bug originates from https://github.com/k3s-io/k3s/blob/9400a8e/pkg/clientaccess/token.go#L232-L235

When the URL verifies with just the system certs, no certs are returned, making it impossible to calculate the correct hash.

**Comment by jandubois** [comment]
Based on @ibuildthecloud's suggestion I tried adding a `--token whatever` argument to avoid calculating the CA hash, but it did not make any difference. k3s startup still loops with

```
INFO[2020-12-19T05:05:35.697744602Z] Waiting for master node  startup: resource name may not be empty
INFO[2020-12-19T05:05:35.851584577Z] Cluster-Http-Server 2020/12/19 05:05:35 http: TLS handshake error from 10.42.0.34:34190: remote error: tls: bad certificate
INFO[2020-12-19T05:05:35.852455810Z] Cluster-Http-Server 2020/12/19 05:05:35 http: TLS handshake error from 10.42.0.34:34192: remote error: tls: bad certificate
INFO[2020-12-19T05:05:36.698282450Z] Waiting for master node  startup: resource name may not be empty
INFO[2020-12-19T05:05:36.853749956Z] Cluster-Http-Server 2020/12/19 05:05:36 http: TLS handshake error from 10.42.0.34:34198: remote error: tls: bad certificate
INFO[2020-12-19T05:05:36.854299609Z] Cluster-Http-Server 2020/12/19 05:05:36 http: TLS handshake error from 10.42.0.34:34196: remote error: tls: bad certificate
ERRO[2020-12-19T05:05:36.990579377Z] token CA hash does not match the server CA hash: d5f671d11474cf600edf08de4bb1e9ecdfb98e5d14ddf97b6033d8aa5376dd4f != e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

**Comment by brandond** [comment]
Note the right-hand hash value it is expecting it to equal is the hash of an empty string:

https://www.di-mgt.com.au/sha_testvectors.html:
>Input message: the empty string "", the bit string of length 0.
>Algorithm	Output
>SHA-256	e3b0c442 98fc1c14 9afbf4c8 996fb924 27ae41e4 649b934c a495991b 7852b855

**Comment by cosmo83** [comment]
Iam having the same exact issue .. Iam trying to copy one of the certs for eirini and placed them in /etc/ssl/certs .  Any quickfix you guys can suggest ?

**Comment by brandond** [comment]
What is eirini?

**Comment by cosmo83** [comment]
https://eirini.cf/

https://kubecf.io/docs/tutorials/deploy-k3s/ (Last couple of lines)

Note on eirini
After deployment if Eirini is enabled, it’s necessary to trust the CA used by eirini to pull images from internal registry on the node.

**Comment by jandubois** [comment]
@mudler You seem to have written the tutorial for KubeCF on k3s; could you verify that the instructions still work?

**Comment by brandond** [comment]
@cosmo83 @jandubois  It appears that the 'bits-service-ssl' secret contains a separate CA cert that needs to be trusted at the host level. This issue is related to issues joining clusters when the Kubernetes cluster CA is trusted. Can you open a different issue for what you're trying to do so as to avoid muddling this one?

**Comment by cosmo83** [comment]
> @mudler You seem to have written the tutorial for KubeCF on k3s; could you verify that the instructions still work?

Works mostly for Kubecf with Diaego, but mostly not with Eirini enabled.

**Comment by rancher-max** [comment]
Saw the error on v1.20.2+k3s1 using the following steps:
1. Bring up node1
2. On a second node run:
```
sudo curl -k https://<node 1 ip>:6443/cacerts -o /usr/local/share/ca-certificates/k3s.crt
sudo update-ca-certificates 
```
3. Join the second node as an agent

The same error reproduces on v1.20.2+k3s1, but is fixed using commit `6e768c301e77738ffd69934363aa0479c4a516d6`
